src/cmdstan/command.hpp

```diff
@@ -197,34 +197,33 @@ int command(int argc, const char *argv[]) {
       diagnostic_json_writers;
   std::vector<stan::callbacks::json_writer<std::ofstream>> metric_json_writers;
 
-  std::string output_base = get_basename_suffix(output_file).first;
-  std::string diagnostic_base;
-  if (!diagnostic_file.empty())
-    diagnostic_base = get_basename_suffix(diagnostic_file).first;
-  else
-    diagnostic_base = output_base;
   bool save_single_paths
       = user_method->arg("pathfinder")
         && (get_arg_val<bool_argument>(parser, "method", "pathfinder",
                                        "save_single_paths"));
-  bool save_diagnostics = diagnostic_base != output_base;
 
   if (user_method->arg("pathfinder")) {
+    if (save_single_paths && diagnostic_file.empty()) {
+      diagnostic_file = output_file;
+    }
+    std::cout << "save_single_paths ? " << save_single_paths
+              << " diagnostic file? " << diagnostic_file << std::endl;
     if (num_chains == 1) {
-      init_filestream_writers(sample_writers, num_chains, id, output_base, "",
+      init_filestream_writers(sample_writers, num_chains, id, output_file, "",
                               ".csv", sig_figs, "# ");
-      if (save_single_paths || save_diagnostics) {
+      if (!diagnostic_file.empty()) {
+        save_single_paths = true;
         init_filestream_writers(diagnostic_json_writers, num_chains, id,
-                                diagnostic_base, "", ".json", sig_figs);
+                                diagnostic_file, "", ".json", sig_figs);
       } else {
         init_null_writers(diagnostic_json_writers, num_chains);
       }
     } else {
-      if (save_single_paths || save_diagnostics) {
-        init_filestream_writers(sample_writers, num_chains, id, output_base,
+      if (save_single_paths) {
```

Comment:
To get exactly the same behavior as before with these arguments, you need to still have an or condition here

```suggestion
      if (save_single_paths || !diagnostic_file.empty()) {
```